src/enlyze/api_clients/base.py

```diff
@@ -0,0 +1,224 @@
+    """Base class for paginated ENLYZE platform API responses using pydantic."""
+
+    data: Any
+
+
+#: TypeVar("M", bound=ApiBaseModel): Type variable serving as a parameter
+# for API response model classes.
+M = TypeVar("M", bound=ApiBaseModel)
+
+
+#: TypeVar("R", bound=PaginatedResponseBaseModel) Type variable serving as a parameter
+# for paginated response models.
+R = TypeVar("R", bound=PaginatedResponseBaseModel)
+
+
+class ApiBaseClient(ABC, Generic[R]):
+    """Client base class encapsulating all interaction with all ENLYZE platform APIs.
+
+    :param token: API token for the ENLYZE platform
+    :param base_url: Base URL of the ENLYZE platform
+    :param timeout: Global timeout for HTTP requests sent to the ENLYZE platform APIs
+
+    """
+
+    PaginatedResponseModel: type[R]
+
+    def __init__(
+        self,
+        token: str,
+        *,
+        base_url: str | httpx.URL = ENLYZE_BASE_URL,
+        timeout: float = HTTPX_TIMEOUT,
+    ):
+        self._client = httpx.Client(
+            auth=TokenAuth(token),
+            base_url=httpx.URL(base_url),
+            timeout=timeout,
+        )
+
+    @cache
+    def _full_url(self, api_path: str) -> str:
+        """Construct full URL from relative URL"""
+        return str(self._client.build_request("", api_path).url)
+
+    def get(self, api_path: str, **kwargs: Any) -> Any:
+        """Wraps :meth:`httpx.Client.get` with defensive error handling
+
+        :param api_path: Relative URL path inside the API name space (or a full URL)
+
+        :raises: :exc:`~enlyze.errors.EnlyzeError` on request failure
+
+        :raises: :exc:`~enlyze.errors.EnlyzeError` on non-2xx status code
+
+        :raises: :exc:`~enlyze.errors.EnlyzeError` on non-JSON payload
+
+        :returns: JSON payload of the response as Python object
+
+        """
+
+        try:
+            response = self._client.get(api_path, **kwargs)
+        except Exception as e:
+            raise EnlyzeError(
+                "Couldn't read from the ENLYZE platform API "
+                f"(GET {self._full_url(api_path)})",
+            ) from e
+
+        try:
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            if e.response.status_code in (
+                HTTPStatus.UNAUTHORIZED,
+                HTTPStatus.FORBIDDEN,
+            ):
+                raise InvalidTokenError
+            else:
+                raise EnlyzeError(
+                    f"ENLYZE platform API returned error {response.status_code}"
+                    f" (GET {self._full_url(api_path)})"
+                ) from e
+
+        try:
+            return response.json()
+        except json.JSONDecodeError as e:
+            raise EnlyzeError(
+                "ENLYZE platform API didn't return a valid JSON object "
+                f"(GET {self._full_url(api_path)})",
+            ) from e
+
+    def _transform_paginated_response_data(self, data: Any) -> Any:
```

Comment:
This, together with `_has_more()` and `_next_page_call_args()` is meant to be overridden by the developer, so kinda part of the outside interface. Should they be consindered private and thus prefix by `_`?

Reply:
They are supposed to be implemented, not overwritten by any client that derives from this class. Moreover, my criterion as to whether a method should be private is whether it's not supposed to be called from outside code. Because none of these methods are supposed to, I went with the underscore prefix.